packages/flame/lib/src/assets/images.dart

```diff
@@ -37,58 +37,73 @@ class Images {
+  /// Adds the [image] into the cache under the key [name].
+  ///
+  /// The cache will assume the ownership of the [image], and will properly
+  /// dispose of it at the end.
   void add(String name, Image image) {
-    _loadedFiles[name] = _ImageAssetLoader(Future.value(image))
-      ..loadedImage = image;
+    _assets[name]?.dispose();
+    _assets[name] = _ImageAsset.fromImage(image);
   }
 
-  /// Remove the image [name] from the cache.
+  /// Removes the image [name] from the cache.
+  ///
+  /// No error is raised if the image [name] is not present in the cache.
   ///
   /// This calls [Image.dispose], so make sure that you don't use the previously
   /// cached image once it is cleared (removed) from the cache.
   void clear(String name) {
-    _loadedFiles.remove(name)?.loadedImage?.dispose();
+    final removedAsset = _assets.remove(name);
+    removedAsset?.dispose();
   }
 
-  /// Remove all cached images.
+  /// Removes all cached images.
   ///
   /// This calls [Image.dispose] for all images in the cache, so make sure that
   /// you don't use any of the previously cached images once [clearCache] has
   /// been called.
   void clearCache() {
-    _loadedFiles.forEach((_, imageAssetLoader) {
-      imageAssetLoader.loadedImage?.dispose();
-    });
-    _loadedFiles.clear();
+    _assets.forEach((_, asset) => asset.dispose());
+    _assets.clear();
   }
 
-  /// Gets the specified image [name] from the cache.
+  /// Returns the image [name] from the cache.
+  ///
+  /// The image returned can be used as long as it remains in the cache, but
+  /// doesn't need to be explicitly disposed.
+  ///
+  /// If you want to retain the image even after you remove it from the cache,
+  /// then you can call `Image.clone()` on it.
   Image fromCache(String name) {
-    final image = _loadedFiles[name];
+    final asset = _assets[name];
     assert(
-      image?.loadedImage != null,
-      'Tried to access a nonexistent entry on cache "$name", make sure to '
-      'use the load method before accessing a file on the cache',
+      asset != null,
+      'Tried to access an image "$name" that does not exist in the cache. Make '
+      'sure to load() an image before accessing it',
     );
-    return image!.loadedImage!;
+    assert(
+      asset!.image != null,
+      'Tried to access an image "$name" before it was loaded. Make sure to '
+      'await the future from load() before using this method',
+    );
+    return asset!.image!;
   }
 
   /// Loads the specified image with [fileName] into the cache.
-  Future<Image> load(String fileName) async {
-    if (!_loadedFiles.containsKey(fileName)) {
-      _loadedFiles[fileName] = _ImageAssetLoader(_fetchToMemory(fileName));
+  Future<Image> load(String fileName) {
+    if (!_assets.containsKey(fileName)) {
+      _assets[fileName] = _ImageAsset.future(_fetchToMemory(fileName));
     }
-    return _loadedFiles[fileName]!.retrieve();
+    return _assets[fileName]!.retrieve();
   }
```

Comment:
This could be rewritten as something like this:
```
FutureOr<Image> load(String filename) async => (_assets[filename] ??= _ImageAsset.future(_fetchToMemory(fileName))).retrieve();
```
But maybe it'd be harder to read.

Reply:
Looks reasonable, I'll try that.